torch/fx/experimental/Partitioner.py

```diff
@@ -0,0 +1,181 @@
+        self.nodes.append(node)
+
+class Partition:
+    """Partition class contains all the information about an individual partition.
+    It also provides necessary methods for manipulation the partition.
+    """
+    def __init__(self, partition_id: int, fx_module: GraphModule) -> None:
+        self.graph_module = fx_module
+        self.nodes: List[Node] = []
+        self.partition_id = partition_id
+        self.parents: List[Partition] = []
+        self.children: List[Partition] = []
+
+    def add_node(self, node: Node) -> None:
+        """Append a new node into the partition."""
+        self.nodes.append(node)
+
+    def add_parent(self, partition: 'Partition') -> None:
+        self.parents.append(partition)
+
+    def add_child(self, partition: 'Partition') -> None:
+        self.children.append(partition)
+
+    def get_children(self) -> List['Partition']:
+        return self.children
+
+    def get_parents(self) -> List['Partition']:
+        return self.parents
+
+    def get_input_nodes(self) -> List[Node]:
+        """Input nodes are coming from two places:
+        placeholder and output from its parents output.
+        """
+        input_nodes: List[Node] = []
+        for node in self.nodes:
+            if node.op == 'placeholder':
+                input_nodes.append(node)
+        for partition in self.parents:
+            input_nodes += partition.get_output_nodes()
+        return input_nodes
+
+    def get_output_nodes(self) -> List[Node]:
+        """Output nodes are the nodes that without any user inside this partition."""
+        output_nodes: List[Node] = []
+        for node in self.nodes:
+            index = self.graph_module.graph.nodes.index(node)
+            user_indexes = GraphManipulation.get_all_users_of(self.graph_module, index)
+            user_nodes = {self.graph_module.graph.nodes[i] for i in user_indexes}
+            # check if user nodes has an intersection with self.nodes
+            if not set(self.nodes).intersection(user_nodes):
+                output_nodes.append(node)
+        return output_nodes
+
+    def __str__(self) -> str:
+        return str(self.partition_id)
+
+class Partitioner:
+    """A graph module may not fit into one device.
+    Partitioner class helps cut one graph into subgraphs (partitions),
+    so that each partition could fit into a different device.
+    """
+    def __init__(self) -> None:
+        """
+        After a partitioner is created,
+        it first check if multiple types of devices (backends) are involved.
+        So far, we assume that there is only one backend and one device with unlimited memory
+        """
+        self.partitions: List[Partition] = []
+        self.devices: List[Dict[str, Union[str, float]]] = []
+        self.node_to_partitions: Dict[Node, List[int]] = {}
+
+    def partition_graph(self, fx_module: GraphModule, devices: List[Dict[str, Union[str, float]]]) -> DAG:
```

Comment:
Consider using [NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) to represent devices